### PR TITLE
Khalil/1888 network config improvements

### PR DESCRIPTION
**Pull request by kc1116**
This PR addresses the current issues related to config management in Flow. Here is the current structure when adding a new config value;
- Add default values
- Add a new CLI flag
- Update node builder BaseConfig
- Update node builder Default BaseConfig
- Update all usages
This process requires multiple changes to multiple files in multiple packages of the code base and can be error-prone when updating or adding multiple config values. This structure leads to default values existing in multiple different places throughout the code base making it hard for developers to get a full picture of the entire Flow configuration with default values. Furthermore, each config value can easily be erroneously overridden by the application code during node startup because the entire NodeConfig is accessible from the node builder struct.

This PR aims to address these issues by providing the following changes;

- [x] Consolidate configuration values into a single, centralized location, eliminating the need for updating multiple files and reducing the likelihood of errors. All configuration values should be initialized in one space from a single source of truth `config.yml` . All configuration values should be accessible from a single package and should not be overridable by the application code.
- [x] Implement a configuration management library [Viper](https://github.com/spf13/viper) to handle the parsing, validation, and merging of configuration files. This allows us to use various config file types. Viper also pairs directly with pflag package allowing CLI flags to easily override default values defined in `config.yml`.
- [x] Adopt a standardized config file for establishing a default configuration
- [x] Reduce the number of steps needed to add a new config value
- [x] Consolidate configuration value validation for all network configuration values.

This PR establishes a new config package and a standardized default configuration file config.yml. config.yml represents the entire Flow config with default values for each of the relevant subcomponents for Flow. You will also find a config.go file that defines the FlowConfig struct. This struct is used to unmarshal config.yml after it has been read. The FlowConfig struct fields are configuration structs for the various subcomponents of Flow. The FlowConfig struct contains a Validate() func that invokes all the validation funcs defined on each sub-configuration structs, allowing devs to define Validate methods of any of the sub-configuration structs. Finally, defining CLI flags for each sub-component should be done in the corresponding sub-package for that component in the config package. Now configuration structs, validation funcs and pflag instantiation are contained in the config sub-package for the component they are related to rather then all flags being instantiated inside scaffold.go separate from config struct definitions and separate from any default value definitions. All configuration for the flow network has been moved to the sub-packge config/network. In the package a configuration struct for each subcomponent of the network is defined, validate funcs are defined as required on any of the configuration structs and all network CLI flags are defined in this package. This setup now contains and updates the entire network configuration from a single package. 

Example steps to add new configuration value for the network config.

- Add new config value to the network-configs key of `config.yml`
```
  networking-connection-pruning: true
```
- Add new field for the config value on the parent network config struct in the config/network/config.go file
```
NetworkConnectionPruning bool `mapstructure:"networking-connection-pruning"`
```
- Add validation for new config value if needed
- Add CLI flag to override defaults to the InitializeNetworkFlags func in config/network/flags.go
```
func InitializeNetworkFlags(flags *pflag.FlagSet, defaultNetConfig *Config) {
	flags.Bool(NetworkingConnectionPruning, defaultNetConfig.NetworkConnectionPruning, "enabling connection trimming")
.  .  .  .  .
}
```
Now adding a new config value and accessing the new config value is completely contained in the config package.

I suggest starting PR review from the new [config package](https://github.com/onflow/flow-go/tree/khalil/1888-network-config-improvements/config), most of the other changes are usage updates.

### CODE REVIEW

**Review by jordanschalm — COMMENTED** [review]
I'm very happy to see movement on standardizing config management - it could certainly use it. Here are some general comments on the approach so far:

### Singleton Pattern
I think we should design the `config` package so that it does not _need_ to be a global singleton. There should a type representing `Config`, a constructor, and getters should be defined on the type, rather than as public methods on a global variable.

### Locality
Most components which can be configured currently define a `Config` type in their local package (and often a default value for these configs). They then accept an instance of `Config` in their constructor. Config validation is also generally implemented in the package. 

We can think of the global app config as the union of these component configs:
```go
type FlowConfig struct {
  NetworkingConfig
  ConsensusEngineConfig
  ...
}
```

I think it is beneficial to maintain this component/config locality, for similar reasons to the above point, so I'm wondering how to incorporate that pattern with the changes you're proposing here. 

One extreme is, the entire app config is passed into to each component. The global `config` package handles dynamic updating, validation etc. And the component just read whichever values they need.

The other extreme is, Viper acts as an initialization process, it produces the initial config as output, from which we can extract component-specific configs, then components are individually responsible for validation, dynamic updating, etc.

I feel like the current approach is leaning toward the first extreme. My gut feeling leans more toward the second extreme, and some of my comments highlight that, but the draft is still early so it's difficult to see. I think applying this approach to one of the components which follow the component-defined config pattern, and teasing out how that will interact with this new package, would be a good next step.

File: network/p2p/utils/ratelimiter/rate_limiter.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ func NewRateLimiter(limit rate.Limit, burst int, lockoutDuration time.Duration,
 		limiterMap:               internal.NewLimiterMap(rateLimiterTTL, cleanUpTickInterval),
 		limit:                    limit,
 		burst:                    burst,
-		rateLimitLockoutDuration: lockoutDuration * time.Second,
+		rateLimitLockoutDuration: lockoutDuration,
```
Comment (peterargue) [inline network/p2p/utils/ratelimiter/rate_limiter.go:42]: can you split these fixes out into their own PR. this looks like the rate limits currently have a permanent lockout

**Review by peterargue — COMMENTED** [review]
Thanks for putting this together @kc1116!

I really like the idea of simplifying the config management, and supporting file based configs. I also think moving to a library makes a lot of sense.

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	configFile embed.FS
+)
+
+// Initialize initializes the flow configuration. All default values for the Flow
+// configuration are stored in the config.yml file. These values can be overriden
+// by node operators by setting the corresponding cli flag. Initialize should be called
+// before any pflags are parsed, this will allow the configuration to initialize with defaults
+// from config.yml.
+func Initialize() error {
+	f, err := configFile.Open(configFileName)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply config/config.go]: https://github.com/onflow/flow-go/pull/4340/commits/3746e223fdce488a71c29c63f51bd091eda7605f

File: config/network.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
+package config
+
+import (
+	"time"
+)
+
+// NetworkConnectionPruning returns the network connection pruning config value.
+func NetworkConnectionPruning() bool {
+	return conf.GetBool(NetworkingConnectionPruningKey)
+}
```
[the remark this answers is not in the document]
Comment (kc1116) [reply config/network.go]: I've taken your feedback into consideration and did away with the singleton style and opted for defining the go structs for each module's config. This updating, validation, and CLI flag instantiation will now be contained the in same package corresponding to the relevant component. i.e: All network configuration is moved to the config/network package.

File: config/network/gossipsub_rpc_inspectors.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	return nil
+}
+
+func (c *CtrlMsgValidationConfig) SetRateLimiter(r p2p.BasicRateLimiter) {
+	c.rateLimiter = r
+}
+
+func (c *CtrlMsgValidationConfig) RateLimiter() p2p.BasicRateLimiter {
+	return c.rateLimiter
+}
```
Comment (jordanschalm) [inline config/network/gossipsub_rpc_inspectors.go:118]: Might make more sense to put this into `ControlMsgValidationInspector`, since it is essentially a dependency of that component. Feels a bit out of place here, since it is created in the constructor, rather than read in from flags/config file. Not a big deal though.

File: follower/follower_builder.go [hunk 5]
```diff
@@ -478,8 +479,9 @@ func (builder *FollowerServiceBuilder) InitIDProviders() {
 			return fmt.Errorf("could not initialize ProtocolStateIDCache: %w", err)
 		}
 		builder.IDTranslator = translator.NewHierarchicalIDTranslator(idCache, translator.NewPublicNetworkIDTranslator())
+		fmt.Println(builder.BaseConfig.FlowConfig)
```
Comment (jordanschalm) [inline follower/follower_builder.go:482]: debug log

File: config/network/errors.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+// ErrInvalidLimitConfig indicates the validation limit is < 0.
+type ErrInvalidLimitConfig struct {
+	// controlMsg the control message type.
+	controlMsg p2p.ControlMessageType
+	err        error
```
Comment (jordanschalm) [inline config/network/errors.go:14]: if we want to be able to interpret this as part of the error chain, need to implement `Unwrap()` method

File: config/network/errors.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+// ErrInvalidLimitConfig indicates the validation limit is < 0.
+type ErrInvalidLimitConfig struct {
```
Comment (jordanschalm) [inline config/network/errors.go:11]:
```suggestion
type InvalidLimitConfigError struct {
```
nitpick: error naming convention

File: config/network/errors.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	// controlMsg the control message type.
+	controlMsg p2p.ControlMessageType
+	err        error
+}
+
+func (e ErrInvalidLimitConfig) Error() string {
+	return fmt.Errorf("invalid rpc control message %s validation limit configuration: %w", e.controlMsg, e.err).Error()
```
Comment (jordanschalm) [inline config/network/errors.go:18]:
```suggestion
	return fmt.Sprintf("invalid rpc control message %s validation limit configuration: %s", e.controlMsg, e.err.Error())
```
not sure we need to construct an error just to get the string representation

File: config/network/errors.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+// NewInvalidLimitConfigErr returns a new ErrValidationLimit.
+func NewInvalidLimitConfigErr(controlMsg p2p.ControlMessageType, err error) ErrInvalidLimitConfig {
+	return ErrInvalidLimitConfig{controlMsg: controlMsg, err: err}
```
Comment (jordanschalm) [inline config/network/errors.go:23]:
```suggestion
func NewInvalidLimitConfigError(controlMsg p2p.ControlMessageType, msg string, args ...any) ErrInvalidLimitConfig {
	return InvalidLimitConfigError{controlMsg: controlMsg, err: fmt.Errorf(msg, args...)}
```

Optional: I find this is a useful pattern to define error types:
- guarantees `MyError.err` is non-nil
- invocation is similar to `fmt.Errorf` + don't need to embed `fmt.Errorf` at the callsite + encourages the caller to add additional context (like `Errorf`)

File: config/config.yml [hunk 10]
```diff
@@ -0,0 +1,124 @@
+# Network Configuration
```
Comment (jordanschalm) [inline config/config.yml:1]: Maybe we could call this file `default-config.yml`?

File: config/config.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/spf13/viper"
+
+	"github.com/onflow/flow-go/config/network"
+)
+
+var (
+	conf = viper.New()
+
+	//go:embed config.yml
+	configFile string
```
Comment (jordanschalm) [inline config/config.go:18]: Right now the yml file essentially acts as an encoded representation of config defaults. Do we want to allow node operators to actually provide a `config.yml` file to configure their node as well? I think doing so would be beneficial (and mostly leverage functionality already provided by Viper). 

But if we support that, we'll need to adjust this package to accept a file input from the node scaffold, and rename `configFile` etc. to make it clear it is only defaults.

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+// by node operators by setting the corresponding cli flag. DefaultConfig should be called
+// before any pflags are parsed, this will allow the configuration to initialize with defaults
+// from config.yml.
+// Returns:
+//
+//	*FlowConfig: an instance of the network configuration fully initialized to the default values set in the config file
+//	error: if there is any error encountered while initializing the configuration, all errors are considered irrecoverable.
+func DefaultConfig() (*FlowConfig, error) {
+	var flowConf FlowConfig
+	err := Unmarshall(&flowConf)
```
Comment (jordanschalm) [inline config/config.go:46]:
```suggestion
	err := conf.Unmarshal(flowConfig, func(decoderConfig *mapstructure.DecoderConfig) {
		// enforce that the default config contains no missing or extraneous fields
		decoderConfig.ErrorUnused = true
		decoderConfig.ErrorUnset = true
	})
```
At least when reading the defaults file (probably not for a user-provided config file, if we choose to support that), checking for missing and extraneous fields would help keep this file neat and tidy.

**Review by jordanschalm — COMMENTED** [review]
> All configuration values should be initialized in one space from a single source of truth config.yml

Have you given any thought to role-specific configs? For example, only the consensus node runs the DKG engine. It defines those configs [within its `main.go` file](https://github.com/onflow/flow-go/blob/1063a5184fd198e2cab60caac361d25aef2ef081/cmd/consensus/main.go#L93-L94), separate from the `FlowNodeBuilder` config fields. 

Would we want to include those role-specific configs in the central config package as well, or is the idea to focus on role-generic configs only?

**Review by AlexHentschel — APPROVED** [review]
Overall, this is a great improvement. I have some questions / worries (details below) regarding flexibility. But even if we don't address those right away, I think we should move ahead with this PR and continue to experiment, iterate and refine with subsequent PRs. 

### High-level thoughts on design and approach

A 'unified framework' on the one hand and 'maximum flexibility' on the other hand are often opposing goals. So far, we were very much on the 'maximum flexibility' side of the spectrum and I agree with all the problems you have explained in your PR comment. However, I think it is important to be careful not to overcorrect towards 'one-size fits all'. 

In particular, I have questions how to address the following scenarios without your framework:
* In my mind, the `flow-go` repo is a collection of libraries (building blocks) to compose various _different_ micro-services (collector, consensus, execution, verification, access, observer, archive node + light clients in the future). 
   * The way I understand your proposal, you are suggesting to have a _single_ default config for everything. I think that introduces multiple fundamental challenges, which I would like to avoid:
   * I consider it very well possible, that a single default configuration satisfying all possible micro-services does not exist. 
   
      For example, consider consensus nodes who have private Random beacon keys that all other nodes do not have. If you configure a default path where those keys are to be found, and that path is verified as part of the `Validate()` method, only the consensus nodes will start and all other nodes will reject the config saying that the given path does not exist. Please only consider this example for illustrative purposed, potentially this is not even a problem right now, or we can work around it.  

   Assuming that there is only a single default configuration is a relatively foundational limitation in my opinion. I would prefer if we had the _option_ to have one default configuration per microservice.

**Review by yhassanzadeh13 — APPROVED** [review]
Thank you for taking the initiative to refactor the default configs, making them cleaner and more maintainable. It looks excellent! I have added a few minor suggestions for cosmetic improvements. Once you’ve made these adjustments, please go ahead and merge. Great work! 

👏

File: config/network/errors.go (hunk 9 above)
Comment (kc1116) [reply config/network/errors.go]: This pattern is less useful for the usage of NewInvalidLimitConfigError since the error is simple the only context needed is the control msg type and the validation err to wrap.

File: config/network/gossipsub_rpc_inspectors.go (hunk 4 above)
Comment (kc1116) [reply config/network/gossipsub_rpc_inspectors.go]: https://github.com/onflow/flow-go/pull/4340/commits/b35ae57f69e529f440036312f964143b3a7cef8b

File: follower/follower_builder.go (hunk 5 above)
Comment (kc1116) [reply follower/follower_builder.go]: https://github.com/onflow/flow-go/pull/4340/commits/412fe62560cf8116ca9f727049e4a2a4e35c5ea9

File: config/network/errors.go (hunk 6 above)
Comment (kc1116) [reply config/network/errors.go]: https://github.com/onflow/flow-go/pull/4340/commits/5f52386b4b348f5f6a88e12c9b902cd80ffcef4f

File: config/config.yml (hunk 10 above)
Comment (kc1116) [reply config/config.yml]: https://github.com/onflow/flow-go/pull/4340/commits/143d8ecefb5db4670f347b832413a8291e0e333e

File: config/config.go (hunk 11 above)
Comment (kc1116) [reply config/config.go]: https://github.com/onflow/flow-go/pull/4340/commits/c353c042a9acd9e5fb0cc315401470789f64dd2a

File: config/config.go (hunk 12 above)
Comment (kc1116) [reply config/config.go]: https://github.com/onflow/flow-go/pull/4340/commits/0371afbcf19887f7aacbd83a78cc164c00c23f13

For now we allow unused keys until all flags and configs are moved to the config package.

**Comment by kc1116** [comment]
> All configuration values should be initialized in one space from a single source of truth config.yml

> Have you given any thought to role-specific configs? For example, only the consensus node runs the DKG engine. It defines those configs [within its `main.go` file](https://github.com/onflow/flow-go/blob/1063a5184fd198e2cab60caac361d25aef2ef081/cmd/consensus/main.go#L93-L94), separate from the `FlowNodeBuilder` config fields. 
 Would we want to include those role-specific configs in the central config package as well, or is the idea to focus on role-generic configs only?

I think we should group configs by component or collection of components (used together and required for some specific functionality). There may or may not be components used across multiple node roles. For example, the "Network" is not an explicitly defined component but a set of components that work together to ensure a functioning network. Furthermore "Consensus" is not an explicitly defined component but a collection of components that work together to ensure a functioning consensus layer. So I would want to include all configuration structs for consensus-related components under a single package in the config package. Currently, DKG configs are defined in the main.go for consensus nodes but it probably should be defined in a config/consensus package where we can take a look at a single package and get a full view of all consensus-related configs.

@jordanschalm

**Comment by kc1116** [comment]
> In my mind, the flow-go repo is a collection of libraries (building blocks) to compose various different micro-services > >(collector, consensus, execution, verification, access, observer, archive node + light clients in the future).
>
> The way I understand your proposal, you are suggesting to have a single default config for everything. I think that introduces multiple fundamental challenges, which I would like to avoid:

> I consider it very well possible, that a single default configuration satisfying all possible micro-services does not exist.
>
> For example, consider consensus nodes who have private Random beacon keys that all other nodes do not have. If you configure a default path where those keys are to be found, and that path is verified as part of the Validate() method, only the consensus nodes will start and all other nodes will reject the config saying that the given path does not exist. Please only consider this example for illustrative purposed, potentially this is not even a problem right now, or we can work around it.
> 
> Assuming that there is only a single default configuration is a relatively foundational limitation in my opinion. I would prefer if we had the option to have one default configuration per microservice.

@AlexHentschel I agree we should not have a single config for each, and thats not an issue. Currently there is a single default-config.yml because only the network-config is defined in the package. In order to improve readability we can define a .yml file per sub config package in the future. The main goal is to move all configuration to a centralized location to improve clarity, and make updating or editing configurations simpler.

File: network/p2p/inspector/internal/ratelimit/control_message_rate_limiter.go [hunk 13]
```diff
@@ -20,6 +21,11 @@ var _ p2p.BasicRateLimiter = (*ControlMessageRateLimiter)(nil)
 // NewControlMessageRateLimiter returns a new ControlMessageRateLimiter. The cleanup loop will be started in a
 // separate goroutine and should be stopped by calling Close.
 func NewControlMessageRateLimiter(limit rate.Limit, burst int) p2p.BasicRateLimiter {
+	if limit == 0 {
+		// setup noop rate limiter if rate limiting is disabled
+		return ratelimit.NewNoopRateLimiter()
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/ratelimit/control_message_rate_limiter.go:28]: https://github.com/onflow/flow-go/pull/4340/commits/ed3b2e844418eca008ec39331f478249e3faa920

File: config/config.go [hunk 14]
```diff
@@ -92,6 +113,71 @@ func Unmarshall(flowConfig *FlowConfig) error {
...
+}
+
+// overrideConfigFile overrides the default config file by reading in the config file at the path set
+// by the --config-file and --config-file-name flags in our viper config store.
+//
+// Returns:
+//
+//	error: if there is any error encountered while reading new config file, all errors are considered irrecoverable.
+//	bool: true if the config was overridden by the new config file, false otherwise or if an error is encountered reading the new config file.
+func overrideConfigFile(flags *pflag.FlagSet) (error, bool) {
```
Comment (yhassanzadeh13) [inline config/config.go:172]: The function `overrideConfigFile` returns `(error, bool)`. It is more idiomatic in Go to have the error as the last return value. So it is better to have it return `(bool, error)`.

File: config/config.go [hunk 15]
```diff
@@ -92,6 +113,71 @@ func Unmarshall(flowConfig *FlowConfig) error {
 	return nil
 }
 
+// Print prints current configuration keys and values.
```
Comment (yhassanzadeh13) [inline config/config.go:129]: Regarding "prints current configuration keys and values", the documentation does not explicitly show any print statements. Instead, it stores key-value pairs into a logger (info). This might be slightly confusing for someone reading the comment and expecting direct print output. To improve the structure of the `Print` function, I'd recommend renaming it to better reflect its purpose and improving its documentation. I'd also recommend passing the necessary configurations as parameters instead of relying on the global state.

Suggestion: 
```go
// LogConfig logs configuration keys and values if they were overridden with a config file.
// It also returns a map of keys for which the values were set by a config file.
//
// Parameters:
//     - logger: *zerolog.Event to which the configuration keys and values will be logged.
//     - flags: *pflag.FlagSet containing the set flags.
//     - conf: configuration store (assumed to be of a type that supports AllKeys and Get methods).
//
// Returns:
//     - map[string]struct{}: map of keys for which the values were set by a config file.
func LogConfig(logger *zerolog.Event, flags *pflag.FlagSet, conf interface{ AllKeys() []string; Get(key string) interface{} }) map[string]struct{} {
	keysToAvoid := make(map[string]struct{})

	if flags.Lookup(configFilePath).Changed {
		for _, key := range conf.AllKeys() {
			logger.Str(key, fmt.Sprintf("%v", conf.Get(key)))
			parts := strings.Split(key, ".")
			if len(parts) == 2 {
				keysToAvoid[parts[1]] = struct{}{}
			} else {
				keysToAvoid[key] = struct{}{}
			}
		}
	}

	return keysToAvoid
}
```

File: config/config.go [hunk 16]
```diff
@@ -92,6 +113,71 @@ func Unmarshall(flowConfig *FlowConfig) error {
...
+		if err != nil {
+			return fmt.Errorf("failed to read config file %s: %w", p, err), false
+		}
+		return nil, true
+	}
+	return nil, false
+}
+
+// getConfigNameFromPath returns the directory and name of the config file from the provided path string.
+func splitConfigPath(path string) (string, string) {
```
Comment (yhassanzadeh13) [inline config/config.go:189]: In this function, you are trying to split the path into the directory and the base name of the config file without the extension. This is fine, but it’s good to note that if someone names a config file with multiple dots (e.g., `my.config.yaml`), this function would consider `my` as the name which might not be intended. I suggest the following implementation that uses a regular expression pattern and panics if the input path does not match the pattern: 

```go
// splitConfigPath returns the directory and base name (without extension) of the config file from the provided path string.
// If the file name does not match the expected pattern, the function panics.
//
// The expected pattern for file names is that they must consist of alphanumeric characters, hyphens, or underscores,
// followed by a single dot and then the extension.
//
// Legitimate Inputs:
//     - /path/to/my_config.yaml
//     - /path/to/my-config123.yaml
//     - my-config.yaml (when in the current directory)
//
// Illegitimate Inputs:
//     - /path/to/my.config.yaml (contains multiple dots)
//     - /path/to/my config.yaml (contains spaces)
//     - /path/to/.config.yaml (does not have a file name before the dot)
//
// Args:
//     - path: The file path string to be split into directory and base name.
//
// Returns:
//     - The directory and base name without extension.
//
// Panics:
//     - If the file name does not match the expected pattern.
func splitConfigPath(path string) (string, string) {
	// Regex to match filenames like 'my_config.yaml' or 'my-config.yaml' but not 'my.config.yaml'
	validFileNamePattern := regexp.MustCompile(`^[a-zA-Z0-9_-]+\.[a-zA-Z0-9]+$`)

	dir, name := filepath.Split(path)

	// Panic if the file name does not match the expected pattern
	if !validFileNamePattern.MatchString(name) {
		panic(fmt.Errorf("Invalid config file name '%s'. Expected pattern: alphanumeric, hyphens, or underscores followed by a single dot and extension", name))
	}

	// Extracting the base name without extension
	baseName := strings.Split(name, ".")[0]
	return dir, baseName
}

```

File: config/network/config.go [hunk 17]
```diff
@@ -62,16 +77,49 @@ type UnicastRateLimitersConfig struct {
...
+			m[s[1]] = key
+		}
+	}
+	// each flag name should correspond to exactly one key in our config store after it is loaded with the default config
+	for _, flagName := range AllFlagNames() {
+		fullKey, ok := m[flagName]
+		if !ok {
+			return fmt.Errorf("invalid network configuration missing configuration key flag name %s check config file and cli flags", flagName)
+		}
+		conf.RegisterAlias(fullKey, flagName)
```
Comment (yhassanzadeh13) [inline config/network/config.go:98]: shouldn't this be `conf.RegisterAlias(flagName, fullKey)` according to the documentation of Viper?

File: config/network/flags.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	ValidationInspectorClusterPrefixedTopicsReceivedCacheDecay = "gossipsub-cluster-prefix-tracker-cache-decay"
+	ValidationInspectorClusterPrefixHardThreshold              = "gossipsub-rpc-cluster-prefixed-hard-threshold"
+
+	// gossipsub metrics inspector
+	MetricsInspectorNumberOfWorkers = "gossipsub-rpc-metrics-inspector-workers"
+	MetricsInspectorCacheSize       = "gossipsub-rpc-metrics-inspector-cache-size"
+
+	ALSPDisabled            = "alsp-disable"
+	ALSPSpamRecordCacheSize = "alsp-spam-record-cache-size"
+	ALSPSpamRecordQueueSize = "alsp-spam-report-queue-size"
```
[the remark this answers is not in the document]
Comment (kc1116) [reply config/network/flags.go]: https://github.com/onflow/flow-go/pull/4340/commits/46f1b7603a7524fe4f7c612367f9be4409561529

File: config/config.go (hunk 14 above)
Comment (kc1116) [reply config/config.go]: https://github.com/onflow/flow-go/pull/4340/commits/6e8c1a1ddfedcb4b68d8b3fd320a2e9e547266f7

File: config/config.go (hunk 15 above)
Comment (kc1116) [reply config/config.go:130]: https://github.com/onflow/flow-go/pull/4340/commits/320c424cb08961c69424af4fac139f5cd362d970

File: config/config.go [hunk 19]
```diff
@@ -92,6 +113,71 @@ func Unmarshall(flowConfig *FlowConfig) error {
...
+// map[string]struct{}: map of keys to avoid printing if they were set by an config file.
+// This is required because we still have other config values not migrated to the config package. When a
+// config file is used currently only network-configs are set, we want to avoid printing the config file
+// value and also the flag value.
+func Print(info *zerolog.Event, flags *pflag.FlagSet) map[string]struct{} {
+	// only print config values if they were overridden with a config file
+	m := make(map[string]struct{})
+	if flags.Lookup(configFilePath).Changed {
+		for _, key := range conf.AllKeys() {
+			info.Str(key, fmt.Sprintf("%v", conf.Get(key)))
```
[the remark this answers is not in the document]
Comment (kc1116) [reply config/config.go:141]: https://github.com/onflow/flow-go/pull/4340/commits/80dac1161c4cbcb19afb8b7494f4542601dd061a

File: config/config.go (hunk 16 above)
Comment (kc1116) [reply config/config.go:241]: https://github.com/onflow/flow-go/pull/4340/commits/99d7a7884da874e9bb274a5a47de30e604d11a15

File: config/network/config.go (hunk 17 above)
Comment (kc1116) [reply config/network/config.go:98]: No the usage is correct. When keys bound from the CLI are added to the config store they are added without any of the prefixes. Because we want to access config values without using long keys with prefixes we set an alias from the keys bound from the CLI -> keys read in from config files that may have prefixes.
